l1-contracts/src/core/Decoder.sol

```diff
@@ -379,7 +429,88 @@ contract Decoder {
+   *      In each iteration of kernel, the kernel computes a hash of the previous iteration's logs hash and the current
+   *      iteration's logs. E.g. for resulting logs hash of a kernel with 3 iterations would be computed as:
+   *
+   *        logsHash = sha256(sha256(sha256(I1_LOGS), I2_LOGS), I3_LOGS)
+   *
+   *      where I1_LOGS, I2_LOGS and I3_LOGS are logs emitted in the first, second and third iterations respectively.
+   *
+   * @dev For the example above, the logs are encoded in the following way:
+   *
+   *        || K_LOGS_LEN | I1_LOGS_LEN | I1_LOGS | I2_LOGS_LEN | I2_LOGS | I3_LOGS_LEN | I3_LOGS ||
+   *           4 bytes      4 bytes       i bytes   4 bytes       j bytes     4 bytes     k bytes
+   *
+   *        K_LOGS_LEN is the total length of the logs in the kernel.
+   *        I1_LOGS_LEN (i) is the length of the logs in the first iteration.
+   *        I1_LOGS are all the logs emitted in the first iteration.
+   *        I2_LOGS_LEN (j) ...
+   *
+   * @dev Link to a relevant discussion:
+   *      https://discourse.aztec.network/t/proposal-forcing-the-sequencer-to-actually-submit-data-to-l1/426/9
+   */
+  function _computeKernelLogsHash(uint256 _offset, bytes calldata /* _l2Block */ )
+    internal
+    pure
+    returns (bytes32, uint256)
+  {
+    uint256 remainingLogsLength;
+    uint256 offset;
+    assembly {
+      // Set the remaining logs length to the total logs length
+      // Loads 32 bytes from calldata, shifts right by 224 bits and masks the result with 0xffffffff
+      remainingLogsLength := and(shr(224, calldataload(_offset)), 0xffffffff)
+      // Move the calldata offset by the 4 bytes we just read
+      offset := add(_offset, 0x4)
+    }
+
+    bytes32 logsHash;
+    uint256 iterationLogsLength;
+    uint256 tempLength;
+
+    // Iterate until all the logs were processed
+    while (remainingLogsLength > 0) {
+      assembly {
+        // Load this iteration's logs length
+        iterationLogsLength := and(shr(224, calldataload(offset)), 0xffffffff)
+        offset := add(offset, 0x4)
+        tempLength := add(0x20, iterationLogsLength) // len(logsHash) + iterationLogsLength
+      }
+
+      // TODO: Allocating memory in each iteration is expensive. Should we somehow set it to max length of all the
+      //       iterations? (e.g. We could do that by first searching for max length in a loop or by modifying
+      //       the encoding and storing max length on a predefined position)
+      bytes memory temp = new bytes(tempLength);
+      assembly {
+        // Copy logsHash from stack to temp
+        mstore(add(temp, 0x20), logsHash)
+
+        // Load this iteration's logs to memory
+        calldatacopy(add(temp, 0x40), offset, iterationLogsLength)
+        offset := add(offset, iterationLogsLength)
+
+        // Compute current iteration's logs hash
+        logsHash := keccak256(add(temp, 0x20), tempLength)
```

Comment:
Why is this keccak? 👀 Will be very expensive to compute in circuits

Reply:
Addressed in [745cf77](https://github.com/AztecProtocol/aztec-packages/pull/615/commits/745cf77eb76c11e4e14c8759479ab5d36375170f)